### PR TITLE
Modified ADS dataset class

### PR DESCRIPTION
**Pull request by lu-ohai**
### Modified ADS dataset class to allow initialize ADSDataset and ADSDatasetWithTarget from constructor and from_dataframe

- Initialize ads dataset instance from `ADSDataset` constructor
- Initialize ads dataset with target from `ADSDatasetWithTarget` constructor
- Build ads dataset instance from `from_dataframe()` of `ADSDataset`
- Build ads dataset with target from `from_dataframe()` of `ADSDatasetWithTarget`

Example
```
# Build ads dataset with target from `from_dataframe()` of `ADSDatasetWithTarget`
employees = ADSDatasetWithTarget.from_dataframe(
    df=pd.read_csv("oci://hosted-ds-datasets@bigdatadatasciencelarge/synthetic/orcl_attrition.csv"),
    target="Attrition",
    storage_options={'config':{},'region':'us-ashburn-1'}
).set_positive_class('Yes')

# Initialize ads dataset with target from `ADSDatasetWithTarget` constructor
employees = ADSDatasetWithTarget(
    df=pd.read_csv("oci://hosted-ds-datasets@bigdatadatasciencelarge/synthetic/orcl_attrition.csv"),
    target="Attrition",
    storage_options={'config':{},'region':'us-ashburn-1'}
)

# Initialize ads dataset instance from `ADSDataset` constructor
employees = ADSDataset(
    df=pd.read_csv("oci://hosted-ds-datasets@bigdatadatasciencelarge/synthetic/orcl_attrition.csv"),
    name="test_dataset",
    description="test_description",
    storage_options={'config':{},'region':'us-ashburn-1'}
)

# Build ads dataset instance from `from_dataframe()` of `ADSDataset`
employees = ADSDataset.from_dataframe(
    df=pd.read_csv("oci://hosted-ds-datasets@bigdatadatasciencelarge/synthetic/orcl_attrition.csv"),
    name="test_dataset",
    description="test_description",
    storage_options={'config':{},'region':'us-ashburn-1'}
)
```

### CODE REVIEW

File: ads/dataset/dataset.py [hunk 1]
```diff
@@ -88,6 +88,17 @@ def __init__(
         # to keep performance high and linear no matter the size of the distributed dataset we
         # create a pandas df that's used internally because this has a fixed upper size.
         #
+        if not isinstance(shape, Tuple):
```
Comment (qiuosier) [inline ads/dataset/dataset.py:91]: `if shape is None` ?

File: ads/dataset/dataset.py [hunk 2]
```diff
@@ -88,6 +88,17 @@ def __init__(
...
+            shape = df.shape
+
+        if not isinstance(sampled_df, pd.DataFrame):
```
Comment (qiuosier) [inline ads/dataset/dataset.py:94]: `if sample_df is None`?

**Review by qiuosier — DISMISSED** [review]

File: ads/dataset/dataset.py (hunk 2 above)
Comment (lu-ohai) [reply ads/dataset/dataset.py:94]: Seems like `pandas.DataFrame` will give ambiguous value error while checking:
Screen Shot 2023-04-28 at 11 46 23 AM
Screen Shot 2023-04-28 at 9 45 50 PM
Using the `isinstance()` to check can avoid both `None` and other object type cases.

Comment (mrDzurb) [reply ads/dataset/dataset.py:94]: is None will not work for DF, however `is not None` should work.

**Review by mrDzurb — APPROVED** [review]

File: ads/dataset/dataset.py (hunk 2 above)
Comment (lu-ohai) [reply ads/dataset/dataset.py]: Thanks @mrDzurb and @qiuosier, just updated

File: docs/source/user_guide/loading_data/connect_legacy.rst [hunk 3]
```diff
@@ -37,14 +38,15 @@ To open a dataset from Object Storage using the resource principal method, you c
   import ads
   import os
 
-  from ads.dataset.factory import DatasetFactory
```
Comment (qiuosier) [inline docs/source/user_guide/loading_data/connect_legacy.rst:40]: `from ads.dataset.dataset import ADSDataset`

File: docs/source/user_guide/loading_data/connect_legacy.rst [hunk 4]
```diff
@@ -148,13 +155,13 @@ After you have stored the ADB username, password, and database name (SID) as var
 
     uri = 'oracle+cx_oracle://' + creds['user'] + ':' + creds['password'] + '@' + creds['sid']
 
-You can use ADS to query a table from your database, and then load that table as an ``ADSDataset`` object through ``DatasetFactory``.
-When you open ``DatasetFactory``, specify the name of the table you want to pull using the ``table`` variable for a given table. For SQL expressions, use the table parameter also. For example, *(`table="SELECT * FROM sh.times WHERE rownum <= 30"`)*.
+You can use ADS to query a table from your database, and then load that table as an ``ADSDatasetWithTarget`` object.
```
Comment (qiuosier) [inline docs/source/user_guide/loading_data/connect_legacy.rst:158]: load that table as a pandas dataframe and convert it to an ``ADSDatasetWithTarget``

File: docs/source/user_guide/loading_data/connect_legacy.rst [hunk 5]
```diff
@@ -269,24 +276,29 @@ You can open Amazon S3 public or private files in ADS. For private files, you mu
...
-  ds = DatasetFactory.open("s3://bucket_name/iris.csv", storage_options = {
+  ds = ADSDataset(
+    df=pd.read_csv("s3://bucket_name/iris.csv"),
+    storage_options = {
       'key': 'aws key',
       'secret': 'aws secret,
       'blocksize': 1000000,
       'client_kwargs': {
-              "endpoint_url": "https://s3-us-west-1.amazonaws.com"
+        "endpoint_url": "https://s3-us-west-1.amazonaws.com"
```
Comment (qiuosier) [inline docs/source/user_guide/loading_data/connect_legacy.rst:286]:
```suggestion
      'endpoint_url': 'https://s3-us-west-1.amazonaws.com'
```

File: docs/source/user_guide/loading_data/connect_legacy.rst [hunk 6]
```diff
@@ -269,24 +276,29 @@ You can open Amazon S3 public or private files in ADS. For private files, you mu
...
       }
   })
 
 
 HTTP(S) Sources
 ===============
 
-To open a dataset from a remote web server source, use ``DatasetFactory.open()`` and specify the URL of the data:
+To open a dataset from a remote web server source, use ``ADSDatasetWithTarget`` and specify the URL of the data:
```
Comment (qiuosier) [inline docs/source/user_guide/loading_data/connect_legacy.rst:294]: use ``pandas`` to load the data from URL and convert it to ``ADSDatasetWithTarget``:

File: docs/source/user_guide/loading_data/connect_legacy.rst [hunk 7]
```diff
@@ -56,19 +58,22 @@ For example:
...
      "profile": "DEFAULT"
-  })
+    }
+  )
 
 Local Storage
 =============
 
-To open a dataset from a local source, use ``DatasetFactory.open`` and specify the path of the data file:
+To open a dataset from a local source, use ``ADSDataset`` and specify the path of the data file:
```
Comment (qiuosier) [inline docs/source/user_guide/loading_data/connect_legacy.rst:72]: use ``pandas`` to load the file into a ``dataframe`` and convert it to ``ADSDataset``:

File: docs/source/user_guide/loading_data/connect_legacy.rst [hunk 8]
```diff
@@ -148,13 +155,13 @@ After you have stored the ADB username, password, and database name (SID) as var
...
-You can use ADS to query a table from your database, and then load that table as an ``ADSDataset`` object through ``DatasetFactory``.
-When you open ``DatasetFactory``, specify the name of the table you want to pull using the ``table`` variable for a given table. For SQL expressions, use the table parameter also. For example, *(`table="SELECT * FROM sh.times WHERE rownum <= 30"`)*.
+You can use ADS to query a table from your database, and then load that table as an ``ADSDatasetWithTarget`` object.
+When you open ``ADSDatasetWithTarget``, specify the name of the table you want to pull using the ``table`` variable for a given table. For SQL expressions, use the table parameter also. For example, *(`table="SELECT * FROM sh.times WHERE rownum <= 30"`)*.
 
 .. code-block:: python3
 
     os.environ['TNS_ADMIN'] = creds['tns_admin']
-    ds = DatasetFactory.open(uri, format="sql", table=table, target='label')
+    ds = ADSDatasetWithTarget(df=pd.read_sql(uri, table=table), target='label')
```
Comment (qiuosier) [inline docs/source/user_guide/loading_data/connect_legacy.rst:164]: This probably does not work? `read_sql` does not seem to take `table`.
https://pandas.pydata.org/docs/reference/api/pandas.read_sql.html

File: docs/source/user_guide/quickstart/quickstart.rst [hunk 9]
```diff
@@ -59,9 +59,9 @@ variable during modeling. The type of this target determines what type of modeli
 to use (regression, binary, and multi-class classification, or time series forecasting).
 
 There are several ways to turn data into an ``ADSDataset``. The simplest way is to
-use `DatasetFactory`, which takes as its first argument as a string URI or a
-``Pandas Dataframe`` object. The URI supports many formats, such as Object Storage
-or S3 files. The
+use `ADSDataset` or `ADSDatasetWithTarget` constructor, which takes as its first argument
+as a ``Pandas Dataframe`` object. The ``Pandas Dataframe`` supports many formats, such as
```
Comment (qiuosier) [inline docs/source/user_guide/quickstart/quickstart.rst:63]: The ``Pandas Dataframe`` supports loading data from many URL schemes.

File: docs/source/user_guide/loading_data/connect_legacy.rst (hunk 8 above)
Comment (lu-ohai) [reply docs/source/user_guide/loading_data/connect_legacy.rst]: Yes, just checked the pandas doc and updated this part. The main change is to first connect and load sql table to pandas and then pass pandas to ads dataset

File: docs/source/user_guide/loading_data/connect_legacy.rst (hunk 3 above)
Comment (lu-ohai) [reply docs/source/user_guide/loading_data/connect_legacy.rst:40]: These are all imported in the very beginning of the doc, and I removed the duplicate imports

**Review by qiuosier — APPROVED** [review]

**Review by z7ye — APPROVED** [review]

**Pull request merged**